src/Microsoft.AspNet.OData.Shared/EdmChangedObjectCollection.cs

```diff
@@ -57,5 +69,288 @@ private void Initialize(IEdmEntityType entityType)
+                        {
+                            if(structuredObj is EdmComplexObject)
+                            {
+                                origStructuredObj = new EdmComplexObject(structuredObj.ActualEdmType as IEdmComplexType);
+                            }
+                            else
+                            {
+                                origStructuredObj = new EdmEntityObject(structuredObj.ActualEdmType as IEdmEntityType);
+                            }
+
+                            originalObj.TrySetPropertyValue(propertyName, origStructuredObj);
+                        }
+
+                        PatchItem(structuredObj, origStructuredObj);                        
+                    }
+                    else
+                    {
+                        originalObj.TrySetPropertyValue(propertyName, value);
+                    }
+                }
+            }
+        }
+
+        private IEdmChangedObject HandleFailedOperation(EdmEntityObject changedObj, DataModificationOperationKind operation, EdmStructuredObject originalObj, 
+            List<IEdmStructuralProperty> keys, string errorMessage)
+        {
+            IEdmChangedObject edmChangedObject = null;
+            DataModificationExceptionType dataModificationExceptionType = new DataModificationExceptionType(operation);
+            dataModificationExceptionType.MessageType = new MessageType();
+            dataModificationExceptionType.MessageType.Message = errorMessage;
+
+            // This handles the Data Modification exception. This adds Core.DataModificationException annotation and also copy other instance annotations.
+            //The failed operation will be based on the protocol
+            switch (operation)
+            {
+                case DataModificationOperationKind.Update:
+                    edmChangedObject = changedObj as IEdmChangedObject;
+                    break;
+                case DataModificationOperationKind.Insert:
+                    {
+                        EdmDeltaDeletedEntityObject edmDeletedObject = new EdmDeltaDeletedEntityObject(EntityType);
+                        PatchItem(edmDeletedObject, changedObj);
+
+                        TryGetContentId(changedObj, keys, edmDeletedObject);
+
+                        edmDeletedObject.TransientInstanceAnnotationContainer = changedObj.TransientInstanceAnnotationContainer;
+                        edmDeletedObject.PersistentInstanceAnnotationsContainer = changedObj.PersistentInstanceAnnotationsContainer;
+
+                        edmDeletedObject.TransientInstanceAnnotationContainer.AddResourceAnnotation("Core.DataModificationException", dataModificationExceptionType);
+                        edmChangedObject = edmDeletedObject;
+                        break;
+                    }
+                case DataModificationOperationKind.Delete:
+                case DataModificationOperationKind.Unlink:
```

Comment:
>Unlink [](start = 51, length = 6)

Do we detect unlink (i.e., a delete with only the id specified)?  it doesn't look like we ever call with operation = DataModificationOperationKind.UnLink, but perhaps we should.

Reply:
We are not handling unlink here right? so can we remove that case?